### PR TITLE
feat: optimize deployment and tally event 

### LINKED ISSUE #937: Use Poseidon libraries which are already deployed
Currently we deploy Poseidon libraries for each MACI deployment which is not efficient. We should instead link them from a registry of already deployed contracts (for each chain).

### LINKED ISSUE #1054: Get rid of poseidon deployment in maci deploy script

### LINKED ISSUE #1060: Poll finalization
# Prerequisites

Please answer the following question for yourself before submitting a feature request.

- [x] I checked to make sure that this request has not already been filed

# Motivation

As a developer, I'd like to check if poll is over and results are calculated.

# Details

Let's add events for poll states and method to check if results are calculated.
 
Make sure that you have reviewed [MACI's Contributor Guidelines](https://github.com/privacy-scaling-explorations/maci/blob/dev/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by 0xmad**
# Description

- [x] Update integration and cli docs
- [x] Add poseidon contract addresses as cli arguments
- [x] Add BallotsTallied(address) event for Tally contract
- [x] Reuse already deployed poseidon contracts
- [x] Use network to distingish contracts for different deployments

## Additional Notes

Storage layout (contracts) and file with contract addresses (cli) are changed

## Related issue(s)

Closes #1054 
Closes #1060
Closes #937 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: contracts/contracts/Tally.sol [hunk 1]
```diff
@@ -135,6 +138,10 @@ contract Tally is Ownable, SnarkCommon, CommonUtilities, Hasher {
     // Update the tally commitment and the tally batch num
     tallyCommitment = _newTallyCommitment;
     tallyBatchNum++;
+
+    if (tallyBatchNum * tallyBatchSize > numSignUps) {
```
Comment (ctrlc03) [inline contracts/contracts/Tally.sol:142]: We can optimize this function by caching the state variable `tallyBatchNum`:

```solidity
    _votingPeriodOver(poll);
    updateSbCommitment();

    // cache the tally batch num
    uint256 _tallyBatchNum = tallyBatchNum;
    // increase it now
    unchecked {
      tallyBatchNum++;
    }

    (, uint256 tallyBatchSize, ) = poll.batchSizes();
    uint256 batchStartIndex = _tallyBatchNum * tallyBatchSize;
    (uint256 numSignUps, ) = poll.numSignUpsAndMessages();

    // Require that there are untalied ballots left
    if (batchStartIndex > numSignUps) {
      revert AllBallotsTallied();
    }

    bool isValid = verifyTallyProof(_proof, numSignUps, batchStartIndex, tallyBatchSize, _newTallyCommitment);
    if (!isValid) {
      revert InvalidTallyVotesProof();
    }

    // Update the tally commitment and the tally batch num
    tallyCommitment = _newTallyCommitment;

    if (_tallyBatchNum + 1 * tallyBatchSize > numSignUps) {
      emit BallotsTallied(address(poll));
    }
```

Comment (ctrlc03) [reply contracts/contracts/Tally.sol:142]: gas used goes from 196232 to 196060 (every little helps)

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Great work, very neat. Just left one suggestion

**Review by kittybest — APPROVED** [review]
lgtm, thanks!

**Review by ctrlc03 — APPROVED** [review]
LGTM thank you!!

**Pull request merged**